clusterloader2/pkg/measurement/util/controlled_pods_indexer.go

```diff
@@ -0,0 +1,135 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	appsinformers "k8s.io/client-go/informers/apps/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	controllerUIDIndex = "controllerUID"
+)
+
+// ControlledPodsIndexer is able to efficiently find pods with ownerReference pointing a given controller object.
+// For Deployments, it performs indirect lookup with ReplicaSets in the middle.
+type ControlledPodsIndexer struct {
+	podsIndexer cache.Indexer
+	podsSynced  cache.InformerSynced
+
+	rsIndexer cache.Indexer
+	rsSynced  cache.InformerSynced
+}
+
+func controllerUIDIndexFunc(obj interface{}) ([]string, error) {
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("object has no meta: %v", err)
+	}
+	controllerRef := metav1.GetControllerOf(meta)
+	if controllerRef == nil {
+		return []string{}, nil
+	}
+	return []string{string(controllerRef.UID)}, nil
+}
+
+// NewControlledPodsIndexer creates a new ControlledPodsIndexer instance.
+func NewControlledPodsIndexer(podsInformer coreinformers.PodInformer, rsInformer appsinformers.ReplicaSetInformer) (*ControlledPodsIndexer, error) {
+	if err := podsInformer.Informer().AddIndexers(cache.Indexers{controllerUIDIndex: controllerUIDIndexFunc}); err != nil {
+		return nil, fmt.Errorf("failed to register indexer: %w", err)
+	}
+	if err := rsInformer.Informer().AddIndexers(cache.Indexers{controllerUIDIndex: controllerUIDIndexFunc}); err != nil {
+		return nil, fmt.Errorf("failed to register indexer: %w", err)
+	}
+
+	return &ControlledPodsIndexer{
+		podsIndexer: podsInformer.Informer().GetIndexer(),
+		podsSynced:  podsInformer.Informer().HasSynced,
+		rsIndexer:   rsInformer.Informer().GetIndexer(),
+		rsSynced:    rsInformer.Informer().HasSynced,
+	}, nil
+}
+
+// WaitForCacheSync waits for all required informers to be initialized.
+func (p *ControlledPodsIndexer) WaitForCacheSync(ctx context.Context) bool {
+	return cache.WaitForNamedCacheSync("PodsIndexer", ctx.Done(), p.podsSynced, p.podsSynced)
+}
+
+// PodsControlledBy returns pods controlled by a given controller object.
+func (p *ControlledPodsIndexer) PodsControlledBy(obj interface{}) ([]*corev1.Pod, error) {
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("object has no meta: %w", err)
+	}
+	typeAccessor, err := meta.TypeAccessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("object has unknown type: %w", err)
+	}
+	res, err := p.appendPodsControlledBy(nil, metaAccessor.GetUID())
```

Comment:
In case of deployments this step is wasted.
Can you instead change it to sth like:
```
... 

var podOwners []string
switch typeAccessor.GetKind() {
case "Deployment":
  replicaSets, err := p.rsIndexer.ByIndex(...)
  if err != nil { ...}
  for _, rs := range replicaSets {
    podOwners = append(podOwners, rs.GetUID())
  }
default:
  podOwners = append(podOwners, metaAccessor.GetUID())
}

for _, owner := range podOwners {
  res, err := appendPodsControllerBy(res, owner)
  if err != nil {...}
}
```

Reply:
Done